File: 513-find-bottom-left-tree-value/513.cpp
#include <iostream>
#include <queue>

using namespace std;

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    int findBottomLeftValue(TreeNode* root) {
        queue<TreeNode*> q;
        q.push(root);
        int res = 0;
        while (!q.empty()) {
            res = q.front() -> val;
            int len = q.size();
            for (int i = 0; i < len; ++i) {
                TreeNode* cur = q.front();
                q.pop();
                if (cur->left != nullptr) {
                    q.push(cur->left);
                }
                if (cur->right != nullptr) {
                    q.push(cur->right);
                }
            }
            if (q.empty()) return res;
        }
        return res;
    }
};

int main(int argc, char** args) {
    Solution sol = Solution();
    TreeNode* t1 = new TreeNode(2);
    TreeNode* t2 = new TreeNode(1);
    TreeNode* t3 = new TreeNode(3);
    
    t1->left = t2;
    t1->right = t3;
    cout << sol.findBottomLeftValue(t1) << endl;

    t1 = new TreeNode(1);
    t2 = new TreeNode(2);
    t3 = new TreeNode(3);
    TreeNode* t4 = new TreeNode(4);
    TreeNode* t5 = new TreeNode(5);
    TreeNode* t6 = new TreeNode(6);
    TreeNode* t7 = new TreeNode(7);
    t1->left = t2;
    t1->right = t3;
    t2->left = t4;
    t3->left = t5;
    t3->right = t6;
    t5->left = t7;
    cout << sol.findBottomLeftValue(t1) << endl;
    
    return 0;
}

File: 513-find-bottom-left-tree-value/513.py
#!/usr/bin/env python
# -*- coding:utf-8 -*-

from collections import deque
from typing import Optional


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def findBottomLeftValue(self, root: TreeNode) -> Optional[int]:
        dq: deque = deque([root])
        while dq:
            tmp = dq[0].val
            for i in range(len(dq)):
                cur = dq.popleft()
                if cur.left:
                    dq.append(cur.left)
                if cur.right:
                    dq.append(cur.right)
            if not dq:
                return tmp


if __name__ == '__main__':
    sol = Solution()
    t1 = TreeNode(2)
    t2 = TreeNode(1)
    t3 = TreeNode(3)
    t1.left = t2
    t1.right = t3
    print(sol.findBottomLeftValue(t1))

    t1 = TreeNode(1)
    t2 = TreeNode(2)
    t3 = TreeNode(3)
    t4 = TreeNode(4)
    t5 = TreeNode(5)
    t6 = TreeNode(6)
    t7 = TreeNode(7)
    t1.left = t2
    t1.right = t3
    t2.left = t4
    t3.left = t5
    t3.right = t6
    t5.left = t7
    print(sol.findBottomLeftValue(t1))
